// @bdsx/image-maps/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLevelStorage = exports.getLevel = exports.getMapData = exports._getLevelStorage = exports.IMAGE_PATH = exports.LevelStorage = void 0;
const bdsx_1 = require("bdsx");
const nativeclass_1 = require("bdsx/nativeclass");
const prochacker_1 = require("bdsx/prochacker");
const core_1 = require("bdsx/core");
const map_data_1 = require("./map-data");
const common_1 = require("bdsx/common");
const level_1 = require("bdsx/bds/level");
const nbt_1 = require("bdsx/bds/nbt");
const path = require("path");
const fs = require("fs");
class LevelStorage extends nativeclass_1.NativeClass {
}
exports.LevelStorage = LevelStorage;
exports.IMAGE_PATH = path.join(process.cwd(), '../images');
if (!fs.existsSync(exports.IMAGE_PATH))
    fs.mkdirSync(exports.IMAGE_PATH);
core_1.pdb.setOptions(common_1.SYMOPT_UNDNAME);
const procHacker = prochacker_1.ProcHacker.load(path.join(__dirname, '../pdb.ini'), [
    "Level::getLevelStorage",
    "MapItemSavedData::setPixel",
    "MapItemSavedData::setLocked",
    "MapItemSavedData::getParentMapId",
    "MapItemSavedData::save"
]);
core_1.pdb.setOptions(0);
const procHackerDec = prochacker_1.ProcHacker.load(path.join(__dirname, '../pdb.ini'), [
    "?getMapSavedData@Level@@UEAAPEAVMapItemSavedData@@AEBVCompoundTag@@@Z"
]);
core_1.pdb.close();
map_data_1.MapItemSavedData.prototype.save = procHacker.js("MapItemSavedData::save", bdsx_1.RawTypeId.Void, { this: map_data_1.MapItemSavedData }, LevelStorage);
map_data_1.MapItemSavedData.prototype.setPixel = procHacker.js("MapItemSavedData::setPixel", bdsx_1.RawTypeId.Void, { this: map_data_1.MapItemSavedData }, bdsx_1.RawTypeId.Int32, bdsx_1.RawTypeId.Int32, bdsx_1.RawTypeId.Int32);
map_data_1.MapItemSavedData.prototype.setLocked = procHacker.js("MapItemSavedData::setLocked", bdsx_1.RawTypeId.Void, { this: map_data_1.MapItemSavedData });
map_data_1.MapItemSavedData.prototype.getMapId = procHacker.js("MapItemSavedData::getParentMapId", bdsx_1.RawTypeId.Bin64, { this: map_data_1.MapItemSavedData });
exports._getLevelStorage = procHacker.js("Level::getLevelStorage", LevelStorage, null, level_1.Level);
exports.getMapData = procHackerDec.js("?getMapSavedData@Level@@UEAAPEAVMapItemSavedData@@AEBVCompoundTag@@@Z", map_data_1.MapItemSavedData, null, level_1.Level, nbt_1.CompoundTag);
function getLevel() {
    return bdsx_1.serverInstance.minecraft.getLevel();
}
exports.getLevel = getLevel;
function getLevelStorage() {
    return exports._getLevelStorage(getLevel());
}
exports.getLevelStorage = getLevelStorage;
bdsx_1.bedrockServer.open.on(() => {
    require('./command');
});
