// @bdsx/image-maps/src/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextFormat = exports.sendMessage = void 0;
const packets_1 = require("bdsx/bds/packets");
function sendMessage(actor, message, type = 1) {
    let pk = packets_1.TextPacket.create();
    pk.type = type;
    pk.message = message;
    pk.sendTo(actor.getNetworkIdentifier());
    pk.dispose();
}
exports.sendMessage = sendMessage;
var TextFormat;
(function (TextFormat) {
    TextFormat["BLACK"] = "\u00A70";
    TextFormat["DARK_BLUE"] = "\u00A71";
    TextFormat["DARK_GREEN"] = "\u00A72";
    TextFormat["DARK_AQUA"] = "\u00A73";
    TextFormat["DARK_RED"] = "\u00A74";
    TextFormat["DARK_PURPLE"] = "\u00A75";
    TextFormat["GOLD"] = "\u00A76";
    TextFormat["GRAY"] = "\u00A77";
    TextFormat["DARK_GRAY"] = "\u00A78";
    TextFormat["BLUE"] = "\u00A79";
    TextFormat["GREEN"] = "\u00A7a";
    TextFormat["AQUA"] = "\u00A7b";
    TextFormat["RED"] = "\u00A7c";
    TextFormat["LIGHT_PURPLE"] = "\u00A7d";
    TextFormat["YELLOW"] = "\u00A7e";
    TextFormat["WHITE"] = "\u00A7f";
    TextFormat["MC_GOLD"] = "\u00A7g";
    TextFormat["OBFUSCATED"] = "\u00A7k";
    TextFormat["BOLD"] = "\u00A7l";
    TextFormat["ITALIC"] = "\u00A7o";
    TextFormat["RESET"] = "\u00A7r";
})(TextFormat = exports.TextFormat || (exports.TextFormat = {}));
